ios/keyboardObserver/REAKeyboardEventObserver.m

```diff
@@ -193,6 +206,15 @@ - (void)recognizeInitialKeyboardState
   });
 }
 
+- (void)clearListeners
+{
+  RCTUnsafeExecuteOnMainQueueSync(^() {
```

Comment:
You don't want listeners to be run after `clearListeners` has been called, that's the root of the issue here.
If you call it asynchronously, after scheduling this the js runtime will be cleaned up, and if the listeners run before this block is executed, you get a crash.
That said, there may be better way of doing this, like having an atomic boolean `invalidated` which stops the listeners from being called - might be worth trying out.

Reply:
I was thinking of other ways to do this. 
The problem with boolean is that at some point before calling the listeners I we have to check if the boolean is set. Something like this:
```
if(!invalidated) {
   for listener in listeners
      call listener
}
```
and now what if we invalidate js runtime *while* calling the listeners? It is still going to crash.
We can try checking the boolean before calling every listener like this:
```
for listener in listeners
   if(!invalidated)
      call listener
```
I didn't test it but I think there is still a couple of instructions between checking the boolean and actually calling a listener so theoretically it can still crash. 

I think there is no way around the fact that js thread has to wait somehow for the main thread to finish calling listeners.

Reply:
The check would have to somehow be done on the JS thread, this way you won't be able to invalidate the runtime while running the listener, since operations on the same thread won't be interleaved. That said, if these listeners run on the main thread, why do they use the JS thread runtime? 
Or am I misunderstanding and the UI thread runtime is destroyed on the JS thread for some reason?

Reply:
According to the xcode debugger it is destroyed on the JS thread. You can even see it on the second screenshot.

Reply:
So the UI thread runtime is destroyed on the JS thread and called on the main thread and we rather can't change that. So I have no other idea than synchronising threads like that.

Reply:
I'd say, the current approach is ok.
I don't think we expect the reload to come from the main thread, so there shouldn't be any deadlocks here.

Actually, does this listener actually run on the JS thread? I'm not sure where `NSNotificationCenter` listeners run.

Reply:
> "Regular notification centers deliver notifications on the thread in which the notification was posted. [...] At times, you may require notifications to be delivered on a particular thread that is determined by you instead of the notification center. [...] In these cases, you must capture the notifications as they are delivered on the default thread and redirect them to the appropriate thread."

So I guess they are run from JS thread then :+1:.